shell/platform/linux/public/flutter_linux/fl_engine.h

```diff
@@ -23,6 +24,99 @@ G_DECLARE_FINAL_TYPE(FlEngine, fl_engine, FL, ENGINE, GObject)
  * #FlEngine is an object that contains a running Flutter engine.
  */
 
+/**
+ * FlEnginePlatformMessageHandle:
+ *
+ * A handle used to respond to platform messages.
+ */
+typedef struct _FlEnginePlatformMessageHandle FlEnginePlatformMessageHandle;
+
+/**
+ * FlEnginePlatformMessageCallback:
+ * @engine: a #FlEngine
+ * @channel: channel message received on.
+ * @message: message content received from Dart
+ * @handle: a handle to respond to the message with
+ * @user_data: (closure): data provided when registering this callback
+ *
+ * Function called when platform messages are received.
+ */
+typedef void (*FlEnginePlatformMessageCallback)(
+    FlEngine* engine,
+    const gchar* channel,
+    GBytes* message,
+    FlEnginePlatformMessageHandle* handle,
+    gpointer user_data);
+
+/**
+ * fl_engine_set_platform_message_handler:
+ * @engine: a #FlEngine
+ * @channel: channel to listen on
+ * @callback: function to call when a message is received on this channel
+ * @user_data: (closure): user data to pass to @callback
+ *
+ * Register a callback to handle platform messages. Call
+ * fl_engine_send_platform_message_response() when this message should be
+ * responded to.
+ */
+void fl_engine_set_platform_message_handler(
+    FlEngine* engine,
+    const gchar* channel,
+    FlEnginePlatformMessageCallback callback,
+    gpointer user_data);
+
+/**
+ * fl_engine_send_platform_message_response:
+ * @engine: a #FlEngine
+ * @handle: message being responded to
+ * @response: (allow-none): response to send or %NULL for an empty response.
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL
+ * to ignore.
+ *
+ * Respond to a platform message.
+ *
+ * Returns: %TRUE on success.
+ */
+gboolean fl_engine_send_platform_message_response(
+    FlEngine* engine,
+    FlEnginePlatformMessageHandle* handle,
+    GBytes* response,
+    GError** error);
+
+/**
+ * fl_engine_send_platform_message:
+ * @engine: a #FlEngine
+ * @channel: channel to send to
+ * @message: (allow-none): message buffer to send or %NULL for an empty message
+ * @cancellable: (allow-none): a #GCancellable or %NULL
```

Comment:
What does this do? It's not clear to me how this interacts with the callback; I can see that they are used to make the task object, but what is it for? (I tried looking at the docs, but didn't see a high-level overview in my initial skim.)

Reply:
GCancellable is the means that the caller can request the the asynchronous call is stopped before the callback has been received. GTask is just a convenience object to make async code easier to write. While we're not able to actually send the cancel through to the Dart code, if a user does cancel the callback will return the error G_IO_ERROR_CANCELLED, which is the expected behaviour and will stop them from processing the response. All GLib async calls should take a GCancellable.